### ISSUE TITLE
Leaking file descriptors when generating too many http requests

### ISSUE THREAD

**Issue by adrai** [aws/aws-sdk-js-v3#3279]
*This issue is extracted from this original issue: https://github.com/aws/aws-sdk-js-v3/issues/3019, like requested: https://github.com/aws/aws-sdk-js-v3/issues/3019#issuecomment-1030267030*

#3019 is concentrated to the readFile (leaking file descriptor) "issue" and this issue is concentrated to the network (leaking file descriptor) "issue"

---

It seems lambda is not "waiting" for the file descriptors to be closed.

This can be observed especially, when having warm lambda executions with a lot of sdk calls, like for DynamoDb or S3, etc...
This each http request opens a network socket which results in an open file descriptor.
Since by default in Node.js the socket timeout is set to 120000ms (2 minutes) it may be the lambda is already finished, but the sockets are still open. When "restarting" the lambda for the next invocations, those file descriptors may still be open.
This leads to this type of EMFILE errors:
```
Error: connect EMFILE 52.94.5.100:443 - Local (undefined:undefined)
Error: getaddrinfo EMFILE dynamodb.eu-west-1.amazonaws.com
Error: A system error occurred: uv_os_homedir returned EMFILE (too many open files)
```

These basic tests shows the count (and leaks) of the emfile count:

Tests originally done in this issue here: https://github.com/aws/aws-sdk-js-v3/issues/3019#issuecomment-1028840006

<details>
<summary>Details</summary>

![image](https://user-images.githubusercontent.com/1086194/152322285-081469dd-6327-4559-819e-bcd9741a8bf2.png)
![image](https://user-images.githubusercontent.com/1086194/152324356-dab148b1-1660-4e7e-9301-0f89a14b50e1.png)

</details>

compared to this tests: https://github.com/aws/aws-sdk-js-v3/issues/3019#issuecomment-1029287130

Defining a custom requestHandler, with a very low socketTimeout reduces drastically the emfiles count:
```js
  requestHandler: new NodeHttpHandler({
      socketTimeout: 10000 // <- this decreases the emfiles count, the Node.js default is 120000
  })
```

<details>
<summary>Details</summary>

![image](https://user-images.githubusercontent.com/1086194/152410670-2a46bea7-83b3-41ff-b582-70d6b7e0ca2f.png)
![image](https://user-images.githubusercontent.com/1086194/152410697-8d1376a6-2945-41b9-bcb3-33051b543ff0.png)

</details>

That's why I suggest to set a low socket Timeout by default, like proposed here: https://github.com/aws/aws-sdk-js-v3/issues/3019#issuecomment-1029713958

proposal:

```js
// https://github.com/aws/aws-sdk-js-v3/blob/main/packages/node-http-handler/src/node-http-handler.ts#L62
socketTimeout: socketTimeout || 10000
```

and probably also here?

```js
// https://github.com/aws/aws-sdk-js-v3/blob/main/packages/node-http-handler/src/node-http2-handler.ts#L44
this.requestTimeout = requestTimeout || 10000;
```

PS. btw. it seems it got worse (more file descriptors) when updating from v3.46.0 to v3.49.0

**Comment by adrai** [comment]
maybe also relevant: https://github.com/aws/aws-sdk-js-v3/issues/2569

**Comment by trivikr** [comment]
The fix in https://github.com/aws/aws-sdk-js-v3/pull/3285 is most likely going to fix this issue too, as the number of file reads is reduced from `16` to `2` for client+operation call.

ToDo: Reproduce this issue using repo https://github.com/samswen/lambda-emfiles when the fix from #3285 is out.

**Comment by adrai** [comment]
This issue (http requests) is also more important for you, right? @ffxsam

**Comment by trivikr** [comment]
> No, I don't think so.

Yup, I'm also not sure and this request will require a deep dive.
That's why I did not mark this issue as fixes, but refs in the PR. And mentioned that we'll try to repro with https://github.com/samswen/lambda-emfiles once fix is out.

> This is not about readFile, but about http requests.

Looking at the bug description, it appears to be because of readFile because of the following reasons:
* EMFILE errors are thrown when there are too many open files.
* It seems it got worse (more file descriptors) when updating from v3.46.0 to v3.49.0, where we added another call to readFile call for defaults.

> That's why you told me to open a new issue.

The reason I asked to create a separate request is for better tracking.

**Comment by adrai** [comment]
fyi: the tests I did where without extra readFile calls, so these are all coming from http requests.
So the tests run with:
```js
import sharedIniFileLoader from '@aws-sdk/shared-ini-file-loader'
Object.assign(sharedIniFileLoader, { loadSharedConfigFiles: async () => ({ configFile: {}, credentialsFile: {} }) })
```

That's why my proposal is to have a very low socketTimeout by default.
In my tests setting a low socketTimeout decreased the EMFILE count from up to over 600 to under 100.

**Comment by trivikr** [comment]
> PS. btw. it seems it got worse (more file descriptors) when updating from v3.46.0 to v3.49.0
> In my tests setting a low socketTimeout decreased the EMFILE count from up to over 600 to under 100 => 6x.

This clearly shows that most of EMFILE errors are caused by file reads being fixed in https://github.com/aws/aws-sdk-js-v3/pull/3285
* Updating from v3.46.0 to v3.49.0 added additional readFile call to read defaults.
* The test setting removes readFile calls for all configurations (except credentials).

The default socketTimeout for the SDK will be application to all environments.
I'll revisit this issue after https://github.com/aws/aws-sdk-js-v3/pull/3285 is out (most likely will be released with v3.51.0 on 02/11)

**Comment by adrai** [comment]
v3.51.0 has been released, any update regarding #3279 / #3291 ?

**Comment by dobrynin** [comment]
Also facing this. Thank for the deep dive @adrai.

**Comment by Muthuveerappanv** [comment]
I'm facing the same issue as well, I was doing 10000 dynamodb updates in 1 chunk and benchmarked the performance with 3.33.x and with 3.49 ran into this EMFILE issues, only option was to reduced the chunk size to 800 (even 1000 didn't work), or rollback to 3.46

For many in production (like me) this is a breaking change. Could we have some updates soon? 

@adrai thank your for the detailed description

**Comment by adrai** [comment]
best advice from my side is really to stay at 3.46 and/or set a lower socketTimeout

**Comment by trivikr** [comment]
@Muthuveerappanv we've reduced readFile calls in https://github.com/aws/aws-sdk-js-v3/pull/3285 which was released in v3.51.0.

Can you test with >3.51.0 and share your results?

**Comment by trivikr** [comment]
@Muthuveerappanv we've reduced readFile calls in https://github.com/aws/aws-sdk-js-v3/pull/3285 which was released in v3.51.0.

Can you test with >3.51.0 and share your results?

**Comment by adrai** [comment]
fyi: with 3.51.0 in my setup, I still get EMFILE errors, too many open sockets... because of the socketTimeout... but it seems it is a very little bit better, but still not good

**Comment by Muthuveerappanv** [comment]
My load test results still hasn't improved, didn't pass even for 1000 parallel dynamodb updates (promise.all), not much improvement [had 10000 updates working fine in 3.33.x, now my chunk size is 800]  > @Muthuveerappanv we've reduced readFile calls in https://github.com/aws/aws-sdk-js-v3/pull/3285 which was released in v3.51.0.
> 
> 
> 
> Can you test with >3.51.0 and share your results?

**Comment by Muthuveerappanv** [comment]
I concur, I tried with version 3.52.0> fyi: with 3.51.0 in my setup, I still get EMFILE errors, too many open sockets... because of the socketTimeout... but it seems it is a very little bit better, but still not good

**Comment by Muthuveerappanv** [comment]
> fyi: with 3.51.0 in my setup, I still get EMFILE errors, too many open sockets... because of the socketTimeout... but it seems it is a very little bit better, but still not good

I concur, I tried with version 3.52.0

**Comment by adrai** [comment]
tldr; the amount of EMFILE generated by readFile is nothing compared to the amount of EMFILE generated by the http requests... that's why you will not notice a decrease of 10 EMFILEs when there are other hundreds or thousands of EMFILEs caused by the http requests

**Comment by Muthuveerappanv** [comment]
> > I concur, I tried with version 3.52.0
> 
> @Muthuveerappanv is the issue in 3.52.0 worse than that in 3.46.0? We think it should be improved as `readFile` calls has reduced from `12` to `2`.

i wouldn't say worse, but definitely hasn't gotten better either. the sdk isn't coping well with the burst http requests made, as it did prior to 3.46.0

**Comment by Muthuveerappanv** [comment]
@trivikr  - any ETA on this bug?

**Comment by trivikr** [comment]
Marking this as guidance, as the issue is specific to high traffic lambdas.
Whenever we visit this issue in future, we need to discuss impact of passing the custom `socketTimeout` value for all environments we support.

**Comment by adrai** [comment]
The question is: Why was this workaround NOT necessary in <= v3.46.0 ?

**Comment by trivikr** [comment]
> Why was this workaround NOT necessary in <= v3.46.0 ?

We suspect that the v3.46.0 added support for defaults mode in https://github.com/aws/aws-sdk-js-v3/pull/3192, which increased the number of readFile calls. I do not see any other change which can cause it from https://github.com/aws/aws-sdk-js-v3/releases/tag/v3.47.0 notes.

If you can provide concrete data for comparisons between, it will assist us to debug this further:
* v3.46.0 which doesn't contain default mode
* v3.47.0 which contains default mode.
* v3.51.0 which reduces readFile calls.

By concrete data, I mean at what number of concurrent requests on Lambda do the EMFILE error occur in these versions.

The previous responses were generic, like the one below:
> i wouldn't say worse, but definitely hasn't gotten better either

**Comment by adrai** [comment]
Here I compared v3.46.0 with v3.49.0: https://github.com/aws/aws-sdk-js-v3/issues/3019#issuecomment-1028840006

**Comment by adrai** [comment]
@ffxsam should have also an example with concurrent S3 requests: https://github.com/aws/aws-sdk-js-v3/issues/3270

**Comment by adrai** [comment]
@trivikr  I think I found what is cause all this extra EMFILEs....

It's probably exactly what @AllanZhengYP commented here: https://github.com/aws/aws-sdk-js-v3/blame/main/packages/node-http-handler/src/node-http-handler.ts#L75

When doing all these hundreds of concurrent requests, the code is not working for the this.config to be ready, and will initialize a loooot of new http(s) clients here: https://github.com/aws/aws-sdk-js-v3/blame/main/packages/node-http-handler/src/node-http-handler.ts#L64

All this was introduced in v3.47.0 with this commit: https://github.com/aws/aws-sdk-js-v3/commit/9152e210c6ec29f34bb070eaf2874039022e6ab7

I tested with this little hack, and it seems to work much better like that:
<img width="999" alt="image" src="https://user-images.githubusercontent.com/1086194/159139905-23e02823-2674-41b6-8c5a-445a0c93894f.png">

**Comment by moltar** [comment]
Ooo nice find @adrai!!

**Comment by adrai** [comment]
@trivikr any updates on this?

**Comment by trivikr** [comment]
@AllanZhengYP will take a look in this issue.

**Comment by kuhe** [comment]
hi @adrai, 

`3.74.0` has been released, which includes a fix based on your screenshot. 

In my test lambda using the code snippet with the looped `.send(new GetObjectCommand(...))` you gave, and wrapped in https://github.com/samswen/lambda-emfiles, the comparison of 3.72.0 and 3.74.0 shows a greatly reduced emfile count, from about 1 per loop iteration to a total of 50, which I believe is consistent with the default maxSockets value in the node http handler config.

**Comment by adrai** [comment]
Your code changes looks good... thank you.
Will test this in the next prod deployment in the next days...
